ml_tools/utils/plotting.py

```diff
@@ -313,40 +320,154 @@ def plot_ice_pdp(models:          Dict[str, PredictionStrategy],
     values = np.random.choice(values, size=min(num_points, len(values)), replace=False)
     values.sort()
 
-    for label, model in models.items():
-        assert model.isTrained, f"Model {label} must be trained before ICE/PDP analysis."
+    chunk_size = max(1, len(values) // num_procs)
+    batches    = [values[i:i + chunk_size] for i in range(0, len(values), chunk_size)]
 
-        ice_data = []
-        for value in values:
+    def process_batch(batch):
+        batch_results = []
+        for value in batch:
             state_series_perturbed = deepcopy(state_series)
             for series in state_series_perturbed:
                 series[state_index][input_feature][input_index] = value
 
             predictions = model.predict(state_series_perturbed)
-            for i, series in enumerate(predictions):
-                ice_data.append({'sample': i,
-                                 'value': value,
-                                 'prediction': series[state_index][output_index]})
 
-            ice_df = pd.DataFrame(ice_data)
+            batch_results.extend([
+                {'sample': i, 'value': value, 'prediction': series[state_index][output_index]}
+                for i, series in enumerate(predictions)
+            ])
+        return batch_results
+
+
+    for label, model in models.items():
+        assert model.isTrained, f"Model {label} must be trained before ICE/PDP analysis."
+
+        if not silent:
+            print(f"Generating ICE/PDP plot: {fig_name_prefix}_{label}_{input_feature}.png")
+            statusbar = StatusBar(len(values))
+
+        ice_data = []
+        with ThreadPoolExecutor(max_workers=num_procs) as executor:
+            jobs = {executor.submit(process_batch, batch): batch for batch in batches}
+
+            completed = 0
+            for job in as_completed(jobs):
+                result = job.result()
+                ice_data.extend(result)
+                if not silent:
+                    batch_size = len(jobs[job])
+                    for _ in range(batch_size):
+                        statusbar.update(completed)
+                        completed += 1
```

Comment:
This also looks the same as that snippet in `state.py`

Reply:
see comment above